server/src/main/java/org/elasticsearch/common/util/Maps.java

```diff
@@ -71,6 +72,26 @@
         return Stream.concat(map.entrySet().stream().filter(k -> key.equals(k.getKey()) == false), Stream.of(entry(key, value)))
                 .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
```

Comment:
possibly we could optimise and return the original map if the key&value are the same as already existing in the original map.